### ISSUE TITLE
Privacy option: Disable storage of IP addresses

### ISSUE THREAD

**Issue by mastuser** [tootsuite/mastodon#6474]
Mastodon appears to log the full user IP addresses (last address in use).

In order to comply with German privacy law and to allow for anonymous use by activists, could you please create an option to disable the storage of IP addresses (possibly making it default)?

Privacy is why many people decide to use Mastodon over Twitter. We wouldn't want a case like [this](http://www.nytimes.com/2011/01/09/world/09wiki.html).

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by Openmedianetwork** [comment]
nice to have an option to keep it for a set time as well as not save it at all.

**Comment by bortzmeyer** [comment]
A case of the police asking data from a Mastodon instance is http://www.laquadrature.net/en/node/10335 (in french only)

**Comment by bortzmeyer** [comment]
Another possibility would be to retain only a prefix (of configurable length) of the IP address. Many programs allow to do so, such as the Web cache Squid.

**Comment by Technowix** [comment]
Keeping a prefix would be a cool compromise :o
It's useful if we want to quickly check for people making multiple accounts in order to harass 😕

**Comment by hellekin** [comment]
People who want to harass on the Web know how to use VPNs or botnets.

**Comment by Technowix** [comment]
So let go those who aren't? 
IP prefix isn't legally able to inculpate someone of something, it just help modération :/

**Comment by hellekin** [comment]
Well, theoretically...

**Comment by bortzmeyer** [comment]
Also, it is not just a matter of law (and the label "legal" is misleading). Above all, it is a political issue: do we want to protect the privacy of the users?

**Comment by mastuser** [comment]
As has been pointed out in the thread linked above, logging IP addresses for 12 months (ip_cleanup_scheduler.rb) is not GDPR compliant (and also politically dangerous for activists). 

I am not sure how to disable the storage of IP addresses altogether. Having ip_cleanup_scheduler.rb clean up ip addresses every time and with no delay would still not prevent them from being stored for a certain amount of time. Also the current_sign_in_ip would still be stored. Maybe removing :trackable from user.rb altogether would work?

I am also not sure whether there is a time limit on IP logging in the session_activations table.

**Comment by rixx** [comment]
Okay, so as far as I can tell, Mastodon keeps only the current and last used IP address, which is already much better than a full log.

Furthermore, Mastodon doesn't seem to *use* those addresses at all – my current workaround of just plain ``update users set last_sign_in_ip=null current_sign_in_ip=null`` doesn't seem to have any negative impact. (Yes, this is not a solution at all, but when used via cronjob it's better than not doing anything to mitigate the situation.)

**Comment by ThomasLeister** [comment]
Me and many other Mastodon admins would very much appreciate if there was an easy way to disable IP address logging / storing. :-) 

Also see: [[German]: Mastodon: Datenschutz und Sicherheit verbesserungsfähig](https://www.kuketz-blog.de/mastodon-datenschutz-und-sicherheit-verbesserungsfaehig/)

**Comment by Perflyst** [comment]
> update users set last_sign_in_ip=null current_sign_in_ip=null

Is this still working? Running this via `bin/update` does not remove IP addresses shown in users security settings.

**Comment by rixx** [comment]
This is an SQL query, not something to be run with ``bin/update``. As it touches the database, please do not just replicate this without understanding what it does.
The complete command to be run as the postgres user is ``/usr/bin/psql mastodon -c "update users set current_sign_in_ip=null, last_sign_in_ip=null; update session_activations set ip=null;"``

**Comment by Perflyst** [comment]
Why do we not just remove the x-real-ip or x-forwarded-for (nginx or apache) in the vhost? This would disable all IP storing, or am I wrong?

**Comment by rixx** [comment]
We'll switch to that option soon – we did not consider it at first because we weren't sure what else might break when removing IP addresses entirely (from caches, etc). Since other administrators have chosen this option, it is now clear that it is safe to do.

**Comment by ichi-i** [comment]
Any news on this?

@rixx have you made that change then?
Did you simply remove these lines:
```
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
```
from the vhost? Was anything else necessary?

**Comment by umonaca** [comment]
Our instance has 2000+ users who do not want their IP addresses to be logged. What we did was just setting one fake IP in the corresponding header on nginx. However, we know that it breaks the rate limit in mastodon for not-logged in users. But as nobody is complaining about hitting rate limit, we decided to keep it as it is for now.

**Comment by noelk** [comment]
this option would be very helpful when running an instance with minimal metadata of users.

**Comment by rriemann-eu** [comment]
Dear all,

the promotion of Mastodon and the fediverse in an environment requiring a high level of data protection compliance is difficult if the processing of logs cannot be configured. I agree that one year is very long and I think we should keep features such as rate limitation per user. What is the status of this issue?

Does anyone figured out a cronjob or alike to remove IP logs older than x days until we have an option in Mastodon?

**Comment by bortzmeyer** [comment]
> Rather late but what about storing the ASN? I feel like that achieves the same purpose as a truncated address while being more effective.

Certainly a very good idea, data minimization. A small practical issue: finding the AS number from the IP address adds complexity (having a copy of the RIS database or a similar one, or querying an external server such as bgp.bortzmeyer.org).

**Comment by kescherCode** [comment]
For instance admins that believe reducing the storage time of IP addresses is a viable alternative for outright disabling, see the changes made in PR #18757, which have been tested on my instance, and are working. The config options may undergo changes before (if ever) getting merged, but you may want to apply this to your own instance, in case this is as much of a pressing issue for you as it was for me.

**Comment by comzeradd** [comment]
We wrote a [sort guide](https://libreops.cc/2022/05/09/mastodon/) on how we currently anonymize IPs in our instance, till there is a better option.

**Comment by kescherCode** [comment]
@comzeradd Making Mastodon think all accesses come from the same IP address may cause users to get rate-limited when trying to log in, effectively creating a DoS vector by disallowing your users to log into your instance.

**Comment by comzeradd** [comment]
Yes, that's a risk but we haven't noticed anything like that for the past few months we are running this. Maybe using `127.0.0.1` is actually having the [reverse effect](https://github.com/mastodon/mastodon/blob/c3aef491d66aec743a3a53e934a494f653745b61/config/initializers/rack_attack.rb#L41-L43)?

**Comment by kescherCode** [comment]
@comzeradd Indeed, I hadn't seen that line before. I guess that is a workaround in case you do not want to collect any IPs at all.

**Comment by tmkis2** [comment]
> We wrote a [sort guide](https://libreops.cc/2022/05/09/mastodon/) on how we currently anonymize IPs in our instance, till there is a better option.

Thank you for this - I modified it a bit for us so we still use the proxy settings but anonymize it. So rate limiting should also be no real problem.

**Comment by AestheticalZ** [comment]
Now that Mastodon is gaining popularity due to the Twitter downfall, wouldn't it make sense to provide an official fix for this after 4 years?

**Comment by leo60228** [comment]
@AestheticalZ That argument applies to every issue. I think the opposite is more true: the devs are likely fairly preoccupied dealing with the massive influx of new users.

**Comment by DatAres37** [comment]
As soon as I start tinkering with the `X-Real-IP` and `X-Forwarded-For` headers I'm regularly getting `404: /api/v1/streaming/` errors in the browser network log. `Subscription error: Unknown stream type` in the mastodon log. 
At the moment I'm adding the header for all websites and then use `proxy_set_header [...] "";` to remove the IP in the mastodon server section. 
I hope there will be a better way to achieve this in the future.

**Comment by yawnbox** [comment]
This is a critical feature to protect the physical safety of all Mastodon users, and minimize the legal liability of admins. Having this on by default would minimize admin's responsibility to fully comply with law enforcement demands. Make Mastodon safer for users by default. Allowing admins to opt-in to collecting this metadata is fine. But defaults matter.

I'm not just giving my feedback randomly. I direct Emerald Onion, i'm experienced with responding to grand jury subpoenas with gag orders for users data (emeraldonion.org/transparency). No one is clearly articulating these risks to new admins of Mastodon.

**Comment by splaGit** [comment]
Please check [this](https://github.com/mastodon/mastodon/blob/main/.env.production.sample).
Admins can disable storage of IPs that easy

**Comment by lukaprincic** [comment]
many who are admining masto instances via managed hosting like masto.host need some kind of feature in the backend without cli access.

**Comment by albjeremias** [comment]
yay, 
so our mastodon instance had troubles this morning during 3h was not connected to any instance. On restart the timeline is not updating anymore.. so I went to try to find out what is happening...

then somehow i ran:
$ docker-compose logs streaming -f 

to check whats the logs in streaming and I got this:
````
streaming_1  | verb Unsubscribe timeline:access_token:7299 
streaming_1  | verb Subscribe timeline:access_token:7299 
streaming_1  | verb xxx-31d0-4f85-8f89-9661245fb9d6 Subscription error: Unknown stream type
streaming_1  | verb Unsubscribe timeline:access_token:7299 
streaming_1  | verb Subscribe timeline:access_token:7299 
streaming_1  | verb xxxx-131f-4cec-bbe9-029c013b7aac Subscription error: Unknown stream type
streaming_1  | verb Unsubscribe timeline:access_token:7299 
streaming_1  | verb Subscribe timeline:access_token:7299 
streaming_1  | verb XXXX-905a-403a-8a9b-750ce16733f4 Subscription error: Unknown stream type
streaming_1  | verb Unsubscribe timeline:access_token:7299 
```

so i search and found this issue...

`Subscription error: Unknown stream type`

I wonder what may cause this error.. on da code i found this:
https://github.com/mastodon/mastodon/blob/8c3c6dcae99c7ce9da7d54adbf7171c8ed8515e9/streaming/index.js#L1028

is it related with misconfiguration of the server? is the database corrupt? should I go for the backup?

**Comment by ThisIsMissEm** [comment]
@albjeremias can you ask this question in the discussion forum, since it's completely unrelated to this issue? https://github.com/mastodon/mastodon/discussions
